test/VarianceTest.cc

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright(c) 2019 Netflix, Inc.
+ * SPDX - License - Identifier: BSD - 2 - Clause - Patent
+ */
+
+/******************************************************************************
+ * @file InvTxfm2dAsmTest.c
+ *
+ * @brief Unit test for variance, mse, sum square functions:
+ * - aom_variance{4-128}x{4-128}_{c,sse2,avx2}
+ * - aom_get_mb_ss_sse2
+ * - aom_mse16x16_{c,avx2}
+ *
+ * @author Cidana-Ryan
+ *
+ ******************************************************************************/
+#include <math.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <new>
+// workaround to eliminate the compiling warning on linux
+// The macro will conflict with definition in gtest.h
+#ifdef __USE_GNU
+#undef __USE_GNU  // defined in EbThreads.h
+#endif
+#ifdef _GNU_SOURCE
+#undef _GNU_SOURCE  // defined in EbThreads.h
+#endif
+#include "aom_dsp_rtcd.h"
+#include "random.h"
+#include "util.h"
+#include "gtest/gtest.h"
+
+using svt_av1_test_tool::SVTRandom;  // to generate the random
+namespace {
+#define MAX_BLOCK_SIZE (128 * 128)
+
+static uint32_t variance_ref(const uint8_t *src, const uint8_t *ref, int width,
+                             int height, int src_stride, int ref_stride,
+                             uint32_t *sse_ptr) {
+    int64_t se = 0;
+    uint64_t sse = 0;
+    for (int y = 0; y < height; y++) {
+        for (int x = 0; x < width; x++) {
+            int diff;
+            diff = src[y * src_stride + x] - ref[y * ref_stride + x];
+            se += diff;
+            sse += diff * diff;
+        }
+    }
+    *sse_ptr = static_cast<uint32_t>(sse);
+    return static_cast<uint32_t>(sse - ((se * se) / (width * height)));
+}
+
+typedef uint32_t (*EB_TEST_MSE_NXM_TYPE)(const uint8_t *src_ptr,
+                                         int32_t src_stride,
+                                         const uint8_t *ref_ptr,
+                                         int32_t ref_stride, uint32_t *sse);
+
+typedef std::tuple<uint32_t, uint32_t, EB_TEST_MSE_NXM_TYPE> MSEParam;
+/**
+ * @brief Unit test mse functions include:
+ *  - aom_mse16x16_{c,avx2}
+ *
+ * Test strategy:
+ *  This test case use random source, ref or max source, zero ref as test
+ * pattern.
+ *
+ *
+ * Expect result:
+ *  Results come from reference functon and target function are
+ * equal.
+ *
+ * Test coverage:
+ *
+ * Test cases:
+ *
+ */
+class MseTest : public ::testing::TestWithParam<MSEParam> {
+  public:
+    MseTest() : width_(TEST_GET_PARAM(0)), height_(TEST_GET_PARAM(1)){};
+    void ref_mse() {
+        int32_t mask = (1 << 8) - 1;
+        SVTRandom *rnd = new SVTRandom(0, mask);
```

Comment:
suggest not new a random tool, just put it on stack by SVTRandom rnd(0, mask); to avoid memory checking and deleting.

Reply:
modified